src/python/pants/backend/python/util_rules/pex_from_targets.py

```diff
@@ -657,5 +663,69 @@ async def get_requirements_pex(request: RequirementsPexRequest, setup: PythonSet
     return pex_request
 
 
+@dataclass(frozen=True)
+class PexReqsRequest:
+    addresses: Addresses
+    interpreter_constraints: InterpreterConstraints | None = None
+
+
+
+@dataclass(frozen=True)
+class PexReqs:
```

Comment:
[Naming is hard. Let's go shopping!](https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Fi.pinimg.com%2Foriginals%2Fb4%2Fca%2Fa1%2Fb4caa15aed00066220110776a113ed29.jpg&f=1&nofb=1)

Reply:
This is all hairy, and I apologize. But I think that the result as it stands might be a bit off. AFAICT, the workflows look like:

* If we are using the new optimization:
    1. Don't actually build a requirements-only PEX, and instead hold onto the requirements which would have been used
    2. When actually creating a PEX containing user code, embed the thirdparty requirements
* If we are not using the optimization:
    1. Build a requirements-only PEX, and put it on the pex-path of another PEX containing user code.

The differentiation between "a requirements PEX" (which contains only thirdparty requirements) and the PEX for user code was intentional (although how important it is in practice is unclear), because thirdparty requirements and your code's transitive thirdparty deps change much less frequently than anything else.

So I think that you'll actually want to preserve that aspect, by adjusting the optimization to always create a `RequirementsPex`, but to do so directly from the Lockfile, rather than from a "repository PEX". That should (AFAICT?) avoid the need for the new type at all, because [the existing `RequirementsPex` creation code](https://github.com/pantsbuild/pants/blob/8d53a4447513de121153a301f737f40679bc6eaa/src/python/pants/backend/python/util_rules/pex_from_targets.py#L624-L625) can be adjusted directly.

Reply:
What's the benefit of building any PEXs at all with the new optimization? AFAICT building PEXs here puts us in the same boat as before w.r.t. maybe wanting `run_against_entire_lockfile`, because for each unique subset of 3rdparty deps we'll be building a PEX and that's non-ideal.

Unless you mean for goals like `package`? I'm pretty sure that code has no different behavior with this PR (although we should probably change that to do bullet point 1, subpoint ii. Today it might be PEXing the whole repo)

Sorry, really confused by this comment :sweat_smile: 

---

> Don't actually build a requirements-only PEX, and instead hold onto the requirements which would have been used
...
>Build a requirements-only PEX, and put it on the pex-path of another PEX containing user code.

That's what this PR is already doing

Reply:
> What's the benefit of building any PEXs at all with the new optimization? AFAICT building PEXs here puts us in the same boat as before w.r.t. maybe wanting `run_against_entire_lockfile`, because for each unique subset of 3rdparty deps we'll be building a PEX and that's non-ideal.

It's not possible to run (a PEX) without building a PEX currently. So in the end, a PEX is built, and then a venv is built from it to actually execute. The primary benefit of building the PEX is that it can be cached: i.e., if you lint or test twice in a row with no code changes, we'll hit the cache for the construction of the PEX, and the only thing that is invalidated is _running_ the built PEX.

Building a requirements/third-party PEX independently means that as long as thirdparty requirements have not changed, you can hit the cache for the network-accessing/thirdparty portion of your build. Without that behavior, changes to _either_ first party or third party code will go through PEX's resolution logic.

To see the difference (and again: note in my comment that "how important it is in practice is unclear"), you'd want to compare the difference between `main` and this branch for a case like: "I ran a test once, made a non-import edit to my test code, and then re-ran". On main, that edit won't actually affect any requirements, and so won't need to hit the network / consume the lockfile at all: on this branch, it will re-run the PEX resolve. Warm runs like this are an important case for local latency, and we need to account for them (even at the expense of CI / cold performance sometimes).

> Sorry, really confused by this comment 😅

Yea, sorry: it's definitely confusing.
 
> > Don't actually build a requirements-only PEX, and instead hold onto the requirements which would have been used
> > ...
> > Build a requirements-only PEX, and put it on the pex-path of another PEX containing user code.
> 
> That's what this PR is already doing

Understood: I was trying to explain the behavior difference between `main` and this branch.

From my perspective, by far the most relevant optimization is skipping the creation of the "repository PEX" (which contains the _entire_ lockfile)... it's not clear to me that adjusting the behavior where we currently create a thirdparty-requirements PEX is necessarily a win.

Reply:
> On main, that edit won't actually affect any requirements, and so won't need to hit the network / run resolution logic / consume the lockfile at all: on this branch, it will re-run the PEX resolve.

I'm not seeing that behavior.

```
joshuacannon@CEPHANDRIUS:~/work/techlabs$ ./pants_from_sources test path/to/a/test_file.py
19:28:19.52 [INFO] Completed: Building pytest.pex from 3rdparty/python/lockfiles/pytest.txt                                                                                 
19:28:19.92 [INFO] Completed: Building local_dists.pex                                                                                                                      
19:28:21.69 [INFO] Completed: Building pytest_runner.pex from 3rdparty/python/lockfiles/repository.lock                                                                     
19:28:22.14 [WARN] Failed to generate JUnit XML data for path/to/a/test_file.py.                                                                                 
19:28:22.14 [ERROR] Completed: Run Pytest - path/to/a/test_file.py failed (exit code 1).
```
Then I touch that file and sprinkle newlines
```
joshuacannon@CEPHANDRIUS:~/work/techlabs$ ./pants_from_sources test path/to/a/test_file.py
19:31:27.77 [WARN] Failed to generate JUnit XML data for path/to/a/test_file.py.                                                                                 
19:31:27.77 [ERROR] Completed: Run Pytest - path/to/a/test_file.py failed (exit code 1).
```

I can share the workunit JSON for proof too.

---

I think maybe the disconnect is we ARE still capturing the third-party reqs separate from first-party user code. `pytest_runner.pex` is a VenvPEX we're building which won't contain the user code. Same for `pylint` and `mypy`.


Let me profile the reqs.pex way to see if my fear about exploding cache or perf is valid. Then we can compare implementations / side-effects.

Reply:
Looks like both solutions have similar characteristics due to PEXs hardlinking/symlinking. So I think the difference between the two is now in the noise and we likely might want to switch gears into what we imagine the future state is, and which option best puts us in that direction.

Reply:
Since the `requirements.pex` way doesn't involve touching client-code, I'm actually leaning that direction. Thanks for the push int he right direction @stuhood